src/honey_badger.rs

```diff
@@ -200,43 +222,213 @@ where
         Ok(())
     }
 
-    /// Checks whether the current epoch has output, and if it does, advances the epoch and
-    /// proposes a new batch.
-    fn process_output(&mut self) -> Result<(), Error> {
-        let old_epoch = self.epoch;
-        while let Some(ser_batches) = self.take_current_output() {
-            // Deserialize the output.
-            let transactions: BTreeMap<NodeUid, Vec<Tx>> = ser_batches
-                .into_iter()
-                .filter_map(|(proposer_id, ser_batch)| {
-                    // If serialization fails, the proposer of that batch is faulty. Ignore it.
-                    bincode::deserialize::<Vec<Tx>>(&ser_batch)
-                        .ok()
-                        .map(|proposed| (proposer_id, proposed))
-                })
-                .collect();
-            let batch = Batch {
-                epoch: self.epoch,
-                transactions,
-            };
+    /// Handles decryption shares sent by `HoneyBadger` instances.
+    fn handle_decryption_share_message(
+        &mut self,
+        sender_id: &NodeUid,
+        epoch: u64,
+        proposer_id: &NodeUid,
+        share: &DecryptionShare<Bls12>,
+    ) -> HoneyBadgerResult<()> {
+        // Insert the share.
+        self.received_shares
+            .entry(epoch)
+            .and_modify(|epoch_entry| {
+                epoch_entry
+                    .entry(proposer_id.clone())
+                    .and_modify(|proposer_entry| {
+                        proposer_entry.insert(sender_id.clone(), share.clone());
+                    })
+                    .or_insert_with(|| {
+                        vec![(sender_id.clone(), share.clone())]
+                            .into_iter()
+                            .collect()
+                    });
+            })
+            .or_insert_with(|| {
+                vec![(
+                    proposer_id.clone(),
+                    vec![(sender_id.clone(), share.clone())]
+                        .into_iter()
+                        .collect(),
+                )].into_iter()
+                    .collect()
+            });
```

Comment:
I think this could be simplified:

```rust
self.received_shares
    .entry(epoch)
    .or_insert_with(BTreeSet::new)
    .entry(proposer_id.clone())
    .or_insert_with(BTreeSet::new)
    .insert(sender_id.clone(), share.clone());
```